### PR TITLE
Initialization dependencies

### PR DESCRIPTION
**Pull request by EMattfolk**

### CODE REVIEW

File: sylt-compiler/src/dependency.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            // namespace in a list for example.
+            if let Read(ident) = &ass.kind {
+                if let Some(Name::Namespace(ns)) =
+                    ctx.compiler.namespaces[ctx.namespace].get(&ident.name)
+                {
+                    deps.insert(*ctx.compiler.namespaces[*ns].get(&field.name).unwrap());
+                    return deps;
+                }
+            }
+            return deps;
```
[the remark this answers is not in the document]
Comment (EMattfolk) [reply sylt-compiler/src/dependency.rs]: I could not figure out how to do it properly. But if it is solved in the compiler, why not solve it here?

File: sylt-compiler/src/dependency.rs [hunk 2]
```diff
@@ -47,29 +50,42 @@ fn assignable_dependencies(ctx: &mut Context, assignable: &Assignable) -> HashSe
...
+                    Read(ident) => {
+                        // Might be shadowed here
+                        let shadowed = ctx.shadowed(&ident.name);
+                        match ctx.compiler.namespaces[ctx.namespace].get(&ident.name) {
+                            Some(&name) if !shadowed => {
+                                let ns = if let Name::Namespace(ns) = name { ns } else { ctx.namespace };
+                                (ns, [name].iter().cloned().collect())
+                            },
+                            _ => (ctx.namespace, HashSet::new()),
+                        }
```
Comment (EMattfolk) [inline sylt-compiler/src/dependency.rs:80]: This code does not work now. I think we need to modify the code a bit.
```
a :: b
b :: A { a: 1 }.a

start :: fn {
    print(a)
}
```

**Review by FredTheDino — APPROVED** [review]
LGTM

File: sylt-compiler/src/dependency.rs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+    fn shadowed(&self, variable: &String) -> bool {
+        return self.variables.iter().rfind(|&v| v == variable).is_some();
+    }
+}
+
+
+
+fn assignable_dependencies(ctx: &mut Context, assignable: &Assignable) -> BTreeSet<Name> {
```
Comment (sornas) [inline sylt-compiler/src/dependency.rs:30]:
```suggestion
}

fn assignable_dependencies(ctx: &mut Context, assignable: &Assignable) -> BTreeSet<Name> {
```

File: sylt-compiler/src/dependency.rs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+use sylt_parser::statement::NameIdentifier;
+
+struct Context<'a> {
+    compiler: &'a Compiler,
+    namespace: usize,
+    variables: Vec<String>,
+}
+
+impl Context<'_> {
+    fn shadow(&mut self, variable: &String) {
```
Comment (sornas) [inline sylt-compiler/src/dependency.rs:17]: Usually &str, not &String.

File: sylt-compiler/src/dependency.rs [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+
+fn assignable_dependencies(ctx: &mut Context, assignable: &Assignable) -> BTreeSet<Name> {
+    use AssignableKind::*;
+    match &assignable.kind {
+        Read(ident) => {
+            match ctx.compiler.namespaces[ctx.namespace].get(&ident.name) {
+                Some(&name) if !ctx.shadowed(&ident.name) => {
+                    [name].iter().cloned().collect()
+                },
```
Comment (sornas) [inline sylt-compiler/src/dependency.rs:37]: I don't really get this but it looks weird.

File: sylt-compiler/src/dependency.rs [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+            .collect(),
+        Expression(expr) => dependencies(ctx, expr),
+    }
+}
+
+fn statement_dependencies(ctx: &mut Context, statement: &Statement) -> BTreeSet<Name> {
+    use StatementKind::*;
+    match &statement.kind {
+        Assignment { target, value, .. } => dependencies(ctx, value)
+            .union(&assignable_dependencies(ctx, target))
```
Comment (sornas) [inline sylt-compiler/src/dependency.rs:101]: Very cool!

File: sylt-compiler/src/dependency.rs [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+            .cloned()
+            .collect(),
+        If { condition, pass, fail } => [
+                dependencies(ctx, condition),
+                statement_dependencies(ctx, pass),
+                statement_dependencies(ctx, fail),
+            ].iter()
+            .flatten()
+            .cloned()
+            .collect(),
```
Comment (sornas) [inline sylt-compiler/src/dependency.rs:111]:
```suggestion
        If { condition, pass, fail } => {
            [
                dependencies(ctx, condition),
                statement_dependencies(ctx, pass),
                statement_dependencies(ctx, fail),
            ]
            .iter()
            .flatten()
            .cloned()
            .collect(),
        }
```

**Review by sornas — APPROVED** [review]

File: sylt-compiler/src/dependency.rs (hunk 5 above)
Comment (EMattfolk) [reply sylt-compiler/src/dependency.rs:37]: It is how you create a set without a variable.

Comment (sornas) [reply sylt-compiler/src/dependency.rs:37]: I see. In that case, I think I'd want it to be a bit more explicit. Maybe `[name.clone].into_iter().collect::<BTreeSet<_>>()` (untested, sorry).

**Pull request merged**